src/rpc/xmlrpc_tinyxml2.cc

```diff
@@ -0,0 +1,451 @@
+    auto value_element_child = value_element->FirstChild();
+    if (value_element_child == nullptr) {
+      return torrent::Object("");
+    }
+    auto base64string = std::string(value_element_child->ToText()->Value());
+    base64string.erase(std::remove_if(base64string.begin(), base64string.end(),
+                                      [](char c) { return c == '\n' || c == '\r'; }),
+                       base64string.end());
+    return torrent::Object(utils::base64decode(base64string));
+  } else {
+    throw xmlrpc_error(XMLRPC_INTERNAL_ERROR, "received unsupported value type: " + value_element_type);
+  }
+  return torrent::Object();
+}
+
+void
+print_object_xml(const torrent::Object& obj, tinyxml2::XMLPrinter* printer) {
+  switch (obj.type()) {
+  case torrent::Object::TYPE_STRING:
+    printer->OpenElement("string", true);
+    printer->PushText(obj.as_string().c_str());
+    printer->CloseElement(true);
+    break;
+  case torrent::Object::TYPE_VALUE:
+    if (obj.as_value() > ((torrent::Object::value_type)2 << 30) || obj.as_value() < -((torrent::Object::value_type)2 << 30)) {
+      printer->OpenElement("i8", true);
+    } else {
+      printer->OpenElement("i4", true);
+    }
+    printer->PushText(std::to_string(obj.as_value()).c_str());
+    printer->CloseElement(true);
+    break;
+  case torrent::Object::TYPE_LIST:
+    printer->OpenElement("array", true);
+    for (const auto& itr : obj.as_list()) {
+      printer->OpenElement("value", true);
+      print_object_xml(itr, printer);
+      printer->CloseElement(true);
+    }
+    printer->CloseElement(true);
+    break;
+  case torrent::Object::TYPE_MAP:
+    printer->OpenElement("struct", true);
+    for (const auto& itr : obj.as_map()) {
+      printer->OpenElement("member", true);
+      printer->OpenElement("name", true);
+      printer->PushText(itr.first.c_str());
+      printer->CloseElement(true);
+      printer->OpenElement("value", true);
+      print_object_xml(itr.second, printer);
+      printer->CloseElement(true);
+      printer->CloseElement(true);
+    }
+    printer->CloseElement(true);
+    break;
+  case torrent::Object::TYPE_DICT_KEY:
+    printer->OpenElement("array", true);
+
+    printer->OpenElement("value", true);
+    print_object_xml(obj.as_dict_key(), printer);
+    printer->CloseElement(true);
+
+    if (obj.as_dict_obj().is_list()) {
+      for (auto itr : obj.as_dict_obj().as_list()) {
+        printer->OpenElement("value", true);
+        print_object_xml(obj.as_dict_key(), printer);
+        printer->CloseElement(true);
+      }
+    } else {
+      printer->OpenElement("value", true);
+      print_object_xml(obj.as_dict_obj(), printer);
+      printer->CloseElement(true);
+    }
+    printer->CloseElement(true);
+    break;
+  default:
+    printer->OpenElement("i4", true);
+    printer->PushText(0);
+    printer->CloseElement(true);
+  }
+}
+
+void
+object_to_target(const torrent::Object& obj, int callFlags, rpc::target_type* target) {
```

Comment:
This should also be used for xmlrpc-c's `xmlrpc_to_target`.

Reply:
I did see that comment, but had deferred on the side of caution since the PR aims to replace that code entirely and there were already concerns about stability.

Reply:
Better to get it done now that it's being worked on, rather than left as a todo that never gets done.

This will also test the code with the xmlrpc-c users.